system/lib/libc/dynlink.c

```diff
@@ -656,12 +656,14 @@ em_promise_t emscripten_dlopen_promise(const char* filename, int flags) {
   // Create a promise that is resolved (and destroyed) once the operation
   // succeeds.
   em_promise_t p = emscripten_promise_create();
-  emscripten_dlopen(filename, flags, p, promise_onsuccess, promise_onerror);
-
   // Create a second promise bound the first one to return the caller.  It's
   // then up to the caller to destroy this promise.
   em_promise_t ret = emscripten_promise_create();
+  
+  // The order matters here. Calling emscripten_dlopen before resolving the second promise
+  // may destroy the first promise before resolving the value.
   emscripten_promise_resolve(ret, EM_PROMISE_MATCH, p);
```

Comment:
When this bug occurs and `emscripten_dlopen` wins the race, IIRC it means that `p` is already resolved by the time we call `emscripten_promise_resolve(ret, EM_PROMISE_MATCH, p)`?

Are you saying that `p` is no longer a valid promise handle at the point?   i.e. the entry in the HandleAllocator map on the JS side is no longer valid?

Or are you saying that `p` is valid but because it already been resolve its somehow too late to link `ret` to it?

Reply:
Well the problem is that `promise_onsuccess` and `promise_onerror` destroy the promise when they are done. If `emscripten_dlopen()` completes synchronously, then it destroys the promise right away. Then `emscripten_promise_resolve()` is too late.